Extensions to log server disk queue (made in the context of version vector/unicast)

sbodagala opened the pull request:
Goal: 

The goal is to hold (version incompatible) disk queue related changes, made in the context of getting recovery to work with version vector, in a branch ("version-vector-disk-queue").

Changes:

Commit 37485d27b47311566ab8a6d0c7f1fc3e3f77e2c9:

- Extend the disk queue entry data structure to hold the following information, needed by the recovery algorithm when unicast is enabled:
- PrevVersion of a version
- List of log servers that a commit proxy sends a commit version to

And, extend the code to populate "unknownCommittedVersions" list (the list of commit versions whose commit status is not known, and to be determined by the recovery version computation algorithm) on a log server restart.

Commit 0b7fbaeea2667d38d2cb8109c2a949d96583c912:

Extend the log server restart logic to read versions from disk queue from known committed version onwards (instead of from the latest known persistent data durable version). The unicast recovery algorithm needs to know all commit versions, from known committed version onwards, that have been received by a log server in order to correctly compute the recovery version.

Commit c6f99e29bec4b64159dceeb0620f13e5f7be958e:

Do not let a log server delete any versions from known committed version onwards from its disk queue.

Commit dfb2d97765d89f2de0e684732c1d822a633b61f1:

Make log servers preserve the disk queue positions of all versions from known committed version onwards (when version vector is enabled).

Commit 9249ca1a01ce0e3e118bec657bd35130d15b8421: 

This change is related to the change made in PR https://github.com/apple/foundationdb/pull/11815 in "main".

In case of spill by reference, log servers should use the logic that is based over "TagData::popped" to decide how long to keep the disk queue positions of versions in memory (instead of using the logic that is based over "LogData::persistentDataVersion", which is applicable to spill by value case).

Commit 5837fd1c58cef0fb6dcc46bfd0b7ae57f71a299f:

Consider all LogData structures while deciding where to start reading the disk queue
from on log server restart.

Commit 16b3aa9579be33a5c05201208726fefe1c67db22:

The known committed version of a LogData structure could be greater than the LogData's version in certain cases (on re-start, because the known committed version could get set based on "revert"). So drop the assertion related to this case.

NOTE: Please note that these changes will cause version incompatibility and so additional code/logic will need to be added to make sure that upgrades/restart related simulation tests work properly.

# Code-Reviewer Section

The general pull request guidelines can be found [here](https://github.com/apple/foundationdb/wiki/FoundationDB-Commit-Process).

Please check each of the following things and check *all* boxes before accepting a PR.

- [ ] The PR has a description, explaining both the problem and the solution.
- [ ] The description mentions which forms of testing were done and the testing seems reasonable.
- [ ] Every function/class/actor that was touched is reasonably well documented.

## For Release-Branches

If this PR is made against a release-branch, please also check the following:

- [ ] This change/bugfix is a cherry-pick from the next younger branch (younger `release-branch` or `main` if this is the youngest branch)
- [ ] There is a good reason why this PR needs to go into a release branch and this reason is documented (either in the description above or in a linked GitHub issue)

sbodagala committed:
- Extend the disk queue entry data structure to hold the following
information, needed by the recovery algorithm when unicast is enabled:
- PrevVersion of a version
- List of log servers that a commit proxy sends a commit version to

And, extend the code to populate "unknownCommittedVersions" list (the
list of commit versions whose commit status is not known, and to be
determined by the recovery version computation algorithm) on a log
server restart.

NOTE: Please note that these changes will cause version incompatibility
and so additional code/logic will need to be added to make sure that
upgrades/restart related simulation tests work properly.

sbodagala committed:
- On restart, make log servers read disk queue entries from known
committed version onwards in the case where the known committed version
is behind LogData::persistentDataDurableVersion (and version vector
unicast is enabled). This is so we will populate "LogData::unknownCommittedVersions",
on log server restart, with all versions that are needed by unicast recovery.

jzhou77 commented on fdbserver/TLogServer.actor.cpp:
```diff
@@ -3425,11 +3471,32 @@ ACTOR Future<Void> restorePersistentState(TLogData* self,
 					if (logData) {
+						if (SERVER_KNOBS->ENABLE_VERSION_VECTOR_TLOG_UNICAST && spillTargetLogDataId.present() &&
+						    qe.id == spillTargetLogDataId.get() && qe.version < logData->persistentDataDurableVersion) {
+							logData->unknownCommittedVersions.emplace_front(qe.version, qe.prevVersion, qe.tLogLocIds);
+
+							while (!logData->unknownCommittedVersions.empty() &&
+							       logData->unknownCommittedVersions.back().version <= logData->knownCommittedVersion) {
+								logData->unknownCommittedVersions.pop_back();
+							}
```
Can you make a function for these lines for code reuse?
There are similar code below and in `tLogCommit()`.

sbodagala replied:
Good point. Done now.

sbodagala committed:
If version vector is enabled then preserve all versions from "knownCommittedVersion"
onwards in log server's disk queue.

sbodagala committed:
- Address a review comment

sbodagala committed:
- Make log servers preserve the disk queue positions of all versions
from known committed version onwards, when version vector is enabled.

sbodagala committed:
- This change is in the context of the change made in PR
https://github.com/apple/foundationdb/pull/11815 in "main".
In case of spill by reference, log servers should use the logic that
is based over "TagData::popped" to decide how long to keep the disk
queue positions of versions in memory (instead of using the logic that
is based over "LogData::persistentDataVersion", which is applicable to
spill by value case).

sbodagala committed:
- Figure out the correct position to start reading the disk queue
from, on log server restart, when unicast is enabled.

sbodagala committed:
- The known committed version of a LogData structure could be greater
than the LogData's version in certain cases (on re-start, because the
known committed version could get set based on "recoverAt"). So drop
the assertion related to this.

dlambrig commented on fdbserver/TLogServer.actor.cpp:
```diff
@@ -1186,7 +1268,9 @@ ACTOR Future<Void> updatePersistentData(TLogData* self, Reference<LogData> logDa
 		if (minVersion != std::numeric_limits<Version>::max()) {
 			self->persistentQueue->forgetBefore(
-			    newPersistentDataVersion,
+			    (!SERVER_KNOBS->ENABLE_VERSION_VECTOR_TLOG_UNICAST
```
[Yes](https://github.com/apple/foundationdb/pull/11815/files)

dlambrig commented on fdbserver/TLogServer.actor.cpp:
```diff
@@ -2430,6 +2514,8 @@ ACTOR Future<Void> tLogCommit(TLogData* self,
 		qe.messages = req.messages;
 		qe.id = logData->logId;
+		qe.prevVersion = req.seqPrevVersion;
+		qe.tLogLocIds = req.tLogLocIds;
```
Here we write the new disk queue entries (prevVersion and tLogLocIds) even when version vector is disabled, which makes sense because the disk format should be the same with and without the feature enabled. We do not populate `tLogLocIds` when version vector is disabled, though. If the version vector knob is changed from disabled->enabled on a machine and during recovery encounter a  `tLogLocIds` that is empty for a version, will that version not be recovered when it should have been?

sbodagala commented on fdbserver/TLogServer.actor.cpp:
```diff
@@ -965,10 +983,24 @@ ACTOR Future<Void> popDiskQueue(TLogData* self, Reference<LogData> logData) {
+	// If version vector is enabled then we need to preserve all versions from "knownCommittedVersion"
+	// onwards (for recovery purpose). Adjust the iterator position accordingly.
+	if (SERVER_KNOBS->ENABLE_VERSION_VECTOR_TLOG_UNICAST &&
+	    logData->knownCommittedVersion < logData->persistentDataVersion) {
```
The scenario could be different in the test I looked at, but I think the change you suggested above would resolve this issue. Please let me know if you want to push this change or you want me to do it, thanks!

dlambrig replied:
can you please do it?

sbodagala replied:
Done, thanks!

sbodagala committed:
- Resolve an issue related to finding the position from which versions
would need to be preserved

sbodagala committed:
- Use the same version that we use in "main" while deciding how long
to keep the disk queue positions of versions in memory (using "minVersion"
here has caused memory issues in performance tests in "main" so do not
use that version here)
